src/Main.elm

```diff
@@ -1,10 +1,132 @@
 module Main exposing (main)
 
-import Data.Audience
-import Data.AudienceFolder
+import Data.Audience exposing ( Audience, subaudiences0 )
+import Data.AudienceFolder exposing ( AudienceFolder, subfolders0, folders0 )
 import Html exposing (Html)
+import Html.Events as Events
+import Html.Attributes as Attr
 
+import Browser
+
+import Dict exposing ( Dict )
+
+-- TODO: remove
+import Debug exposing ( log, todo, toString )
+
+-- ====== Overview ======
+-- 0. To compile the app, execute
+--      $ elm make src/Main.elm --output=public/main.js
+--    and open public/index.html. The css is in public/style.css and is not included when the app is executed via elm reactor.
+-- 1. Data.Audience and Data.AudienceFolder expose dictionaries
+--    whose keys are folder ids and whose values are folders, subfolders, and subaudience of the mentioned folder id.
+-- 2. If all parsing of JSON goes well, the state of the application is just
+--    a triple (record) of parent folder, its subfolders and subaudiences.
+-- 3. The only permissible action is to change the current parent folder.
+
+-- ===STATE=== --
+type alias ExplicitFolder =
+  { parentFolder : AudienceFolder
+  , subfolders : List AudienceFolder
+  , subaudiences : List Audience }
+
+type alias State = Maybe ExplicitFolder
+
+-- a helper function - not worth to put into a separate module
+listOfMaybeList : Maybe (List a) -> List a
+listOfMaybeList mxs =
+  case mxs of
+    Just xs -> xs
+    Nothing -> []
+
+stateOfFolderId : Int -> State
+stateOfFolderId folderId =
+  case (folders0, subfolders0, subaudiences0) of
+    (Ok foldersDict, Ok subfoldersDict, Ok subaudiencesDict) ->
```

Comment:
Interesting use of a precomputation (I guess if we were receiving the data dynamically in runtime, these precomputed dicts would live in Model as a sort of cache?). Out of curiosity, what approach would you choose if the amount of data would be so big that doing this eagerly for all items would be too slow (app unresponsive for x seconds)?

Reply:
Yes, I basically precomputed `folders0, subfolders0, subaudiences0` just by going through all of the json blob.

You mean if the json blob was huge?
In that case, I would need to develop some sort of lazy parsing mechanisms to extract only the data from the json blob that I needed at the moment. But this would depend a lot on how the json blob is structured. For example if I wanted to compute all `subaudiences` of a given folder with `folderId0 : Int`, I would have to depend on some data locality, meaning I would have to rely on something like all audiences with the parent `folderId0` would be in one place in the json blob. If they were scattered all over the place, then this approach would not gain a significant speedup.

I think I would like a more specific question with explicit assumptions about the data to really answer this.

I had to deal with similar concerns in https://omedusyo.github.io/combinatorics/# . Here if I chose `n = 20, m = 10`, and did it in a naive way, I would have to compute an array of length `184756`. I resolved this with laziness via es6 generators.

Reply:
@omedusyo

> You mean if the json blob was huge?
Yes, but let me clarify: for this thought exercise I'm not worried about the duration of parsing JSON, but the preprocessing that you do afterwards.

> I think I would like a more specific question with explicit assumptions about the data to really answer this.
Fair enough :slightly_smiling_face: In our production codebase we have entities `Question` and `Category` (which are roughly similar in hierarchy to `Audience` and `AudienceFolder`: Categories contain Questions), and the JSON blob can sometimes be around 6 MB in size. If you try figuring out all this `Dict CategoryId (List QuestionId, List CategoryId)` parentage stuff eagerly, the page freezes for a few seconds (at least it did when we last tried this). So we had to do something else. I'm interested in what approach other than "do all the preprocessing needed for `view` when receiving the JSON response in `update`" would you try in that situation.

Reply:
The freezing problem can be solved as follows:
I could do the heavy computation of collecting the data in a separate thread (via Web Workers - I don't know how convenient this is in Elm). While the thread is doing its thing, the state of the app would be something like `SearchingRelevantData` and the view would reflect this. Since the computation is in a separate thread, the main thread would not be frozen.

Now I'll try to think through the question more carefuly.
**Assumptions**: I have a 6 MB `data : String` - the json blob. Since you are not worried about the parsing, I assume that I've parsed the blob into `questions : List Question`, and `categories : List Category`. I also have a `question_id : QuestionId`.

**Goal**: Compute the view that displays the subquestions and subcategories of `question_id`.

Now you say that if I tried to precompute everything into the dictionary that maps category id's to the list of subquestions and subcategories - which amounts to iterating through `questions` and `categories` independently () - then this would freeze the page for too long. I have a feeling that this is about O(n log(n)) amount of work. Well... more than linear, but probably less than O(n log(n)).

But I only need to compute relevant data to `question_id` - this is only linear. I could memoize this operation so when I search for the same id later, it would be constant time. But this seems strange to me - I feel like this shouldn't make the difference between freezing and not-freezing. If the precomputation had quadratic time, then that would be believable :smile: I feel like the parsing would freeze the page too. So I don't really know how you did it :open_mouth:

Reply:
Another possibility would be to have the data sorted on the server by the question id's ahead of time so I can get it precomputed.